synapse/util/ratelimitutils.py

```diff
@@ -178,7 +249,10 @@ def should_sleep(self) -> bool:
         return len(self.request_times) > self.sleep_limit
 
     async def _on_enter_with_tracing(self, request_id: object) -> None:
-        with start_active_span("ratelimit wait"), queue_wait_timer.time():
+        maybe_metrics_cm = contextlib.nullcontext()
+        if self.metrics_name:
+            maybe_metrics_cm = queue_wait_timer.labels(self.metrics_name).time()
+        with start_active_span("ratelimit wait"), maybe_metrics_cm:
```

Comment:
Slightly worried about [`context` issues](https://github.com/matrix-org/synapse/pull/13554) here again 😰

I've made `metrics_name` optional because I'm not concerned about tracking the `UsernameAvailabilityRestServlet` which uses it to rate limit people by IP. I only care about tracking the federation servlets which `FederationRateLimiter` was made for.

We could make `metrics_name` required and record metrics for the username availability servlet to get rid of this complexity though.

Reply:
Re-commented on the diff: https://github.com/matrix-org/synapse/pull/13649#discussion_r956335553